### ISSUE TITLE
Consider loading task assemblies using Assembly.Load(byte[])

### ISSUE THREAD

**Issue by KirillOsenkov** [microsoft/msbuild#5385]
A common problem is MSBuild node processes locking a task assembly file on disk and lingering, thus preventing the deletion or update of the file. 

Consider reading the raw bytes from disk and loading using that, to avoid holding the .dll locked. Need to investigate whether there will be performance or other implications.

https://source.dot.net/#System.Private.CoreLib/Assembly.cs,b2d43e31d338b43c

To me the locked task .dll happens often enough that it would be warranted.

**Comment by rainersigwald** [comment]
I think there would be loading-of-dependency implications. Right now those come from the task's directory.

The best workaround for the locking problem at the moment is to specify to run the task in an isolated process with `UsingTask TaskFactory="TaskHostFactory"`. That has perf implications due to IPC overhead, of course. But I don't think there's any no-compromises approach in .NET.

**Comment by KirillOsenkov** [comment]
What if we just flip nodereuse to false by default? How much are we really gaining by reusing nodes?

**Comment by csharper2010** [comment]
> What if we just flip nodereuse to false by default? How much are we really gaining by reusing nodes?

I fully agree with that. We ran into this because we had different versions of a build task assembly in different branches. Building one branch after another lead to usage of wrong version because of reusing a node where the assembly was already loaded.

We finally found out that disabling node reuse is the right choice for us but why on earth is the default for this "ON"? You can't assume that node reuse is safe at least if there are custom build task assemblies. This contradicts reproducable builds.

Or is there a way to define rules about node reuse so the new build knows it can't use nodes from a previous build in another branch?

**Comment by rainersigwald** [comment]
> Or is there a way to define rules about node reuse so the new build knows it can't use nodes from a previous build in another branch?

Specify `TaskFactory="TaskHostFactory"` in the `UsingTask` that references your task. Then it will be run in a fresh process that is never reused across builds.
